### ISSUE TITLE
[Q] What's the status to support Web Stream ?

### ISSUE THREAD

**Issue by AlwaysNoobCoder** [exceljs/exceljs#2753]
Web Stream are supported in modern browsers.

Do we have plan to support it ?  Are there technical obstacles to implement?

see: #1228, after four years it's still open, no response from team members.

**Comment by danielegarciav** [comment]
Might just have to create a fork that uses the Web Stream API instead of Node.js streams.

**Comment by lionel-rowe** [comment]
Here's what I'm using.

<details>
<summary>Code</summary>

```ts
import type { Writable } from 'node:stream'
import type ExcelJS from 'exceljs'

type ReadableStreamResult<T> = {
    readable: ReadableStream<T>
    done: Promise<void>
}

export function workbookToStream(workbook: ExcelJS.Workbook): ReadableStreamResult<Uint8Array> {
    const transform = new TransformStream<Uint8Array, Uint8Array>()
    const done = workbook.xlsx.write(new NodeJsWritableLike<Uint8Array>(transform.writable) as unknown as Writable)

    return {
        readable: transform.readable,
        done,
    }
}

class NodeJsWritableLike<T> {
    writer: WritableStreamDefaultWriter<T>

    constructor(writable: WritableStream<T>) {
        this.writer = writable.getWriter()
    }

    async write(chunk: T, cb: () => void): Promise<void> {
        await this.writer.ready
        await this.writer.write(chunk)
        cb()
    }
    async end(): Promise<void> {
        await this.writer.ready
        await this.writer.close()
    }
}
```
</details>

<details>
<summary>Explanation</summary>

`NodeJsWritableLike` is a minimal wrapper around `WritableStream` that implements only the specific methods and signatures of `node:stream`'s `Writable` that are called internally by `workbook.xlsx.write`.

1. `stream` is passed as an argument to `zip.pipe`:

https://github.com/exceljs/exceljs/blob/5bed18b45e824f409b08456b59b87430ded023ab/lib/xlsx/xlsx.js#L720-L724

2. It's passed as `destination` to `ZipWriter`'s `this.stream`:

https://github.com/exceljs/exceljs/blob/5bed18b45e824f409b08456b59b87430ded023ab/lib/utils/zip-stream.js#L66-L68

3. `ZipWriter`'s `this.stream` is a `StreamBuf`:

https://github.com/exceljs/exceljs/blob/5bed18b45e824f409b08456b59b87430ded023ab/lib/utils/zip-stream.js#L22

4. Upon calling `StreamBuf`'s `pipe` method, it's added to `StreamBuf`'s `this.pipes`:

https://github.com/exceljs/exceljs/blob/5bed18b45e824f409b08456b59b87430ded023ab/lib/utils/stream-buf.js#L342-L344

5. `this.pipes` are only iterated in two places, the first to `write` with a callback:

https://github.com/exceljs/exceljs/blob/5bed18b45e824f409b08456b59b87430ded023ab/lib/utils/stream-buf.js#L202-L211

6. ...and the second to `end` with no callback:

https://github.com/exceljs/exceljs/blob/5bed18b45e824f409b08456b59b87430ded023ab/lib/utils/stream-buf.js#L289-L291
</details>

Seems to work well in both browser (Chrome) and server-side (Deno).

Assuming my logic is sound, it should be foolproof, but not necessarily _future_-proof as the ExcelJS's implementation details could change. I might submit a PR to try and get this properly supported and tested, but for now the wrapper-based method should be good to use as of v4.4.0.

**Comment by markedwards** [comment]
@lionel-rowe there is a [Writable.fromWeb()](https://nodejs.org/api/stream.html#streamwritablefromwebwritablestream-options) method that takes care of this, actually.  It's marked as experimental, but it seems to work well.

**Comment by lionel-rowe** [comment]
> [@lionel-rowe](https://github.com/lionel-rowe) there is a [Writable.fromWeb()](https://nodejs.org/api/stream.html#streamwritablefromwebwritablestream-options) method that takes care of this, actually. It's marked as experimental, but it seems to work well.

Yeah I saw that, but I'd rather not make NodeJS writable streams into a runtime dependency when only a small fraction of their logic is used.

**Comment by markedwards** [comment]
Isn't Node.js `Stream` already a runtime dependency of ExcelJS?

**Comment by lionel-rowe** [comment]
> Isn't Node.js Stream already a runtime dependency of ExcelJS?

I think (?) readable is but not writable, unless writable gets pulled in somewhere indirectly.
